TreasureData.agent.lib.nut

```diff
@@ -0,0 +1,144 @@
+    //                               the "value" is the value to be written into that field.
+    //     callback : function       Optional callback function executed once the operation
+    //         (optional)            is completed.
+    //                               The callback signature:
+    //                               callback(error, data), where
+    //                                   error : table               Error details, HTTP response table (see https://electricimp.com/docs/api/httprequest/sendasync/)
+    //                                                               or null if the operation succeeds.
+    //                                   data : table                The original data passed to 
+    //                                                               sendData() method.
+    function sendData(dbName, tableName, data, callback = null) {
+        _processRequest(
+            "POST",
+            _TREASURE_DATA_SEND_ENDPOINT,
+            format(_TREASURE_DATA_SEND_PATH, dbName, tableName),
+            _sendHeaders,
+            data,
+            function (error) {
+                _invokeDefaultCallback(callback, error, data);
+            }.bindenv(this));
+    }
+
+    // Enables/disables the library debug output (including errors logging).
+    // Disabled by default.
+    //
+    // Parameters:
+    //     value : boolean             true to enable, false to disable
+    function setDebug(value) {
+        _debug = value;
+    }
+
+    // -------------------- PRIVATE METHODS -------------------- //
+
+    // Sends an http request to TreasureData.
+    function _processRequest(method, endpoint, path, headers, body, callback) {
+        local url = endpoint + path;
+        local encodedBody = http.jsonencode(body);
+        _logDebug(format("Doing the request: %s %s, body: %s", method, url, encodedBody));
+
+        local request = http.request(method, url, headers, encodedBody);
+        request.sendasync(function (response) {
+            _processResponse(response, callback);
+        }.bindenv(this));
+    }
+
+    // Processes http response from TreasureData and executes callback.
+    function _processResponse(response, callback) {
+        _logDebug(format("Response status: %d, body: %s", response.statuscode, response.body));
+        local error = null;
+        local httpStatus = response.statuscode;
+        if (httpStatus < 200 || httpStatus >= 300) {
+            error = response;
+        }
+        callback(error);
+    }
+
+    // Logs error occurred and invokes callback with default parameters.
+    function _invokeDefaultCallback(callback, error, data) {
+        if (error) {
+            _logError(format("%s: %i", _TREASURE_DATA_REQUEST_FAILED, error.statuscode));
+        }
+        if (callback) {
+            imp.wakeup(0, function () {
+                callback(error, data);
```

Comment:
I have not used this api, so please let me know if I understand this correctly - 
If an http request is successful the library doesn't expose the http response to the the user only the original data that was written.  
If there was an error the user gets the response back

This feels odd to me - Is there really no useful info in the http response if the request was successful?

Reply:
Yes, your understanding is correct.

In case of no error, http response is empty (only the status is present).
If a response includes some valuable info (eg. for a GET request) it will be parsed and returned via an additional meaningful parameter.

So, here "httpresponse" is used as an error indication only. In many other libs we have a Error class for that. As this lib is very primitive so far, Pavel suggested to eliminate the Error class and use "httpresponse" instead of it (I personally do not like this idea very much, but it's Ok).

Reply:
Great if the response body is empty than this seems fine.